kt/godot-runtime/src/main/kotlin/godot/core/GarbageCollector.kt

```diff
@@ -0,0 +1,143 @@
+package godot.core
+
+import godot.util.VoidPtr
+import java.lang.ref.WeakReference
+import java.time.Duration
+import java.time.Instant
+
+object GarbageCollector : Thread() {
+    val refWrappedMap = mutableMapOf<VoidPtr, WeakReference<KtObject>>()
+    val wrappedMap = mutableMapOf<VoidPtr, KtObject>()
+    val refWrappedSuppressBuffer = mutableListOf<VoidPtr>()
+    val wrappedSuppressBuffer = mutableListOf<VoidPtr>()
+
+    private var forceJvmGarbageCollector = false
+    private var gcState = GCState.NONE
+    val isClosed: Boolean
+        get() = gcState == GCState.CLOSED
+
+    fun registerInstance(instance: KtObject) {
+        synchronized(this) {

```

Comment:
Point of contention, do we need to lock every time or only when we start the gc checks?

Reply:
Well we have to lock the container each time we want to read/write. Otherwise we take the risk of deleting/adding stuff at the same time.
But I have plans to add queue for adding and removing stuff. That way we will only lock on the queue instead of the whole class.